inject/service/src/main/java/io/helidon/inject/service/ServiceDescriptor.java

```diff
@@ -0,0 +1,111 @@
+    /**
+     * Create a new service instance.
+     *
+     * @param ctx                  injection context with all injection points data
+     * @param interceptionMetadata interception metadata to use when the constructor should be intercepted
+     * @return a new instance, must be of the type T or a subclass
+     */
+    // we cannot return T, as it does not allow us to correctly handle inheritance
+    default Object instantiate(InjectionContext ctx, InterceptionMetadata interceptionMetadata) {
+        throw new IllegalStateException("Cannot instantiate type " + serviceType().fqName() + ", as it is either abstract,"
+                                                + " or an interface.");
+    }
+
+    /**
+     * Inject fields and methods.
+     *
+     * @param ctx                  injection context
+     * @param interceptionMetadata interception metadata to support interception of field injection
+     * @param injected             mutable set of already injected methods from subtypes
+     * @param instance             instance to update
+     */
+
+    default void inject(InjectionContext ctx,
+                        InterceptionMetadata interceptionMetadata,
+                        Set<String> injected,
+                        T instance) {
+    }
+
+    /**
+     * Invoke {@link Injection.PostConstruct} annotated method(s).
+     *
+     * @param instance instance to use
+     */
+    default void postConstruct(T instance) {
+    }
+
+    /**
+     * Invoke {@link Injection.PreDestroy} annotated method(s).
+     *
+     * @param instance instance to use
+     */
+    default void preDestroy(T instance) {
+    }
+
+    /**
+     * Combine dependencies from this type with dependencies from supertype.
+     * This is a utility for code generated types.
+     *
+     * @param myType    this type's dependencies
+     * @param superType super type's dependencies
+     * @return a new list without constructor dependencies from super type
+     */
+    default List<Ip> combineDependencies(List<Ip> myType, List<Ip> superType) {
+        List<Ip> result = new ArrayList<>(myType);
+
+        // always inject all fields
+        result.addAll(superType.stream()
+                              .filter(it -> it.elementKind() == ElementKind.FIELD)
+                              .toList());
+        // ignore constructors, as we only need to inject constructor on the instantiated type
+
+        // and only add methods that are not already injected on existing type
+        Set<String> injectedMethods = myType.stream()
+                .filter(it -> it.elementKind() == ElementKind.METHOD)
+                .map(Ip::method)
+                .flatMap(Optional::stream)
+                .collect(Collectors.toSet());
+
+        result.addAll(superType.stream()
+                              .filter(it -> it.elementKind() == ElementKind.METHOD)
+                              .filter(it -> injectedMethods.add(it.method().orElse("")))
```

Comment:
Suspicious, IP with `elementType==METHOD` can have empty `.method()`?

Reply:
because I do not like calling `.get()` on `Optional` when not checked

Reply:
Technically, you could use `.orElseThrow()`